hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/ratis/OzoneManagerDoubleBuffer.java

```diff
@@ -0,0 +1,212 @@
+            ExitUtil.terminate(1, message);
+          }
+        });
+
+        try {
+          omMetadataManager.getStore().commitBatchOperation(batchOperation);
+        } catch (IOException ex) {
+          // During flush to rocksdb got an exception.
+          // We should terminate the OM.
+          String message = "During flush to DB encountered error " +
+              ex.getMessage();
+          ExitUtil.terminate(1, message);
+        }
+
+        int flushedTransactionsSize = readyBuffer.size();
+        flushedTransactionCount.addAndGet(flushedTransactionsSize);
+        flushIterations.incrementAndGet();
+
+        LOG.info("Sync Iteration {} flushed transactions in this iteration{}",
+            flushIterations.get(), flushedTransactionsSize);
+        readyBuffer.clear();
+        syncToDB = false;
+        // TODO: update the last updated index in OzoneManagerStateMachine.
+      }
+    }
+  }
+
+  /**
+   * Returns the flushed transaction count to OM DB.
+   * @return flushedTransactionCount
+   */
+  public long getFlushedTransactionCount() {
+    return flushedTransactionCount.get();
+  }
+
+  /**
+   * Returns total number of flush iterations run by sync thread.
+   * @return flushIterations
+   */
+  public long getFlushIterations() {
+    return flushIterations.get();
+  }
+
+  /**
+   * Add OmResponseBufferEntry to buffer.
+   * @param response
+   * @param transactionIndex
+   */
+  public synchronized void add(OMClientResponse response,
+      long transactionIndex) {
+    currentBuffer.add(new DoubleBufferEntry<>(transactionIndex, response));
+  }
+
+  /**
+   * Check can we flush transactions or not.
+   *
+   * @return boolean
+   */
+  @SuppressFBWarnings(value="IS2_INCONSISTENT_SYNC", justification =
+      "Just checking size here, so we don't need synchronize here")
+  private boolean canFlush() {
+    return !syncToDB && currentBuffer.size() > 0;
+  }
+
+  /**
+   * Prepares the readyBuffer which is used by sync thread to flush
+   * transactions to OM DB.
+   */
+  private synchronized void setReadyBuffer() {
+    TransactionBuffer temp = currentBuffer;
+    currentBuffer = readyBuffer;
+    readyBuffer = temp;
+  }
+
+  /**
+   * TransactionBuffer which holds queue of
+   * {@link org.apache.hadoop.ozone.om.ratis.helpers.DoubleBufferEntry}.
+   */
+  static class TransactionBuffer {
+
+    private Queue<DoubleBufferEntry<OMClientResponse>> queue;
+
+    TransactionBuffer() {
+      // Taken unbounded queue, if sync thread is taking too long time, we
+      // might end up taking huge memory to add entries to the buffer.
+      // TODO: We can avoid this using unbounded queue and use queue with
+      // capcity, if queue is full we can wait for sync to be completed to
+      // add entries. But in this also we might block rpc handlers, as we
+      // clear entries after sync. Or we can come up with a good approach to
+      // solve this.
+      queue = new ConcurrentLinkedQueue<>();
```

Comment:
Why not make this bounded and have a force sync function like edits double buffer did? Unbounded queue is uncontrollable, we might pass the init buffer size as a bounded value.

Reply:
So your suggestion is in add check if it is full have a flush() and then add to buffer?

Reply:
We can setReadyBuffer if we check the buffer if full, so this will not block requests long time. Then background will flush ready buffer. Only when the ready buffer is flushing meanwhile current buffer is full, the request will be blocked.

Reply:
If flush happens in background, requests will not be blocked right?

One disadvantage I see is if flush is taking a too long time, and currentBuffer can grow very high and may cause OOM. So, if we use a fixed size buffer we might not see but we will block further transactions.

Reply:
I can see the tradeoffs of both approaches. Why don't benchmark this and see the maximum length this queue grows under load?. One suggestion based on @linyiqun thought is that it would be good to have a counter to see the max length of this queue. So we can at least identify and track this issue if it really pops up in the field.

Reply:
The unbounded nature will be a problem only if the disk is really slow i.e. flush call blocks. Since we are always having a flush operation outstanding.

So the tradeoff is unbounded queue vs blocking RPC handlers. I am okay with unbounded queue to start with.